apps/emqx_resource/src/emqx_resource_worker.erl

```diff
@@ -608,32 +760,24 @@ maybe_append_queue(Id, Index, Q, Items) ->
                 ?SLOG(error, #{msg => drop_query, reason => queue_full, dropped => Dropped}),
                 Q1
         end,
+    Items = [?Q_ITEM(X) || X <- Queries],
     Q3 = replayq:append(Q2, Items),
     emqx_resource_metrics:queuing_set(Id, Index, replayq:count(Q3)),
+    ?tp(resource_worker_appended_to_queue, #{id => Id, items => Queries}),
     Q3.
 
+-spec get_first_n_from_queue(replayq:q(), pos_integer()) ->
+    empty | {replayq:q(), replayq:ack_ref(), [?Q_ITEM(?QUERY(_From, _Request, _HasBeenSent))]}.
 get_first_n_from_queue(Q, N) ->
-    get_first_n_from_queue(Q, N, []).
-
-get_first_n_from_queue(_Q, 0, Acc) ->
-    lists:reverse(Acc);
-get_first_n_from_queue(Q, N, Acc) when N > 0 ->
-    case replayq:peek(Q) of
```

Comment:
Changed to pop because, previously, the head of the queue would be duplicated N times.